src/docfx/lib/schema/JsonSchemaTransform.cs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class JsonSchemaTransform
+    {
+        public static (List<Error> errors, JToken token) Transform(Document file, Context context, JsonSchema schema, JToken token, Action<Document> buildChild)
+        {
+            var errors = new List<Error>();
+
+            var transformedToken = JsonUtility.DeepClone(token);
+            Transform(file, context, schema, transformedToken, errors, buildChild);
+            return (errors, transformedToken);
+        }
+
+        private static void Transform(Document file, Context context, JsonSchema schema, JToken token, List<Error> errors, Action<Document> buildChild)
+        {
+            switch (token)
+            {
+                case JValue scalar:
+                    token.Replace(TransformScalar(file, context, schema, scalar, errors, buildChild));
+                    break;
+                case JArray array:
+                    foreach (var a in array)
+                    {
+                        Transform(file, context, schema, a, errors, buildChild);
+                    }
+                    break;
+                case JObject obj:
+                    foreach (var (key, value) in obj)
+                    {
+                        if (schema.Properties.TryGetValue(key, out var propertySchema))
+                        {
+                            Transform(file, context, propertySchema, value, errors, buildChild);
+                        }
+                    }
+                    break;
+            }
+        }
+
+        private static JValue TransformScalar(Document file, Context context, JsonSchema schema, JValue value, List<Error> errors, Action<Document> buildChild)
+        {
+            if (value.Type == JTokenType.Null)
+            {
+                return value;
+            }
+
+            var dependencyResolver = file.Schema.Type == typeof(LandingData) ? context.LandingPageDependencyResolver : context.DependencyResolver;
+            var sourceInfo = JsonUtility.GetSourceInfo(value);
+            var content = new SourceInfo<string>(value.Value<string>(), sourceInfo);
+
+            switch (schema.ContentType)
+            {
+                case JsonSchemaContentType.Href:
+                    var (error, link, _) = dependencyResolver.ResolveLink(content, file, file, buildChild);
+                    errors.AddIfNotNull(error);
+                    content = new SourceInfo<string>(link, content);
```

Comment:
Can directly return here, source info will not be useful afterwards (some transforms like xref does not return a simple string)

```suggestion
                    return link;
```

Reply:
i would like to keep the original source info in case we use it somewhere someday.